src/crxviewer.js

```diff
@@ -887,6 +895,73 @@ var viewFileInfo = (function() {
     return viewFileInfo;
 })();
 
+function parseExtensionFromManifest(manifest) {
+    var id;
+    if (manifest.browser_specific_settings &&
+        manifest.browser_specific_settings.gecko &&
+        (id = manifest.browser_specific_settings.gecko.id)) {
+        return id;
+    }
+
+    if (manifest.applications && manifest.applications.gecko &&
+        (id = manifest.applications.gecko.id)) {
+        return id;
+    }
+
+    return '';
+}
+
+// Extract the extension ID from the currently loaded extension. The ID will be
+// determined from the COSE signature, the PKCS#7 signature, or manifest.json,
+// whatever comes first. Resolves to the extension ID (and the extension type
+// between parentheses if available), or rejects on parser errors.
+function calculateExtensionID() {
+    var fileList = document.getElementById('file-list');
+    var readEntry = function(filename, Writer) {
+        var li = fileList.querySelector('li[data-filename="' + filename + '"]');
+        if (!li) {
+            return null;
+        }
+        return new Promise(function(resolve) {
+            li.zipEntry.getData(new Writer(), function(data) {
+                resolve(data);
+            });
+        });
+    };
+    var formatSubject = function(subject) {
+        return subject.CN + ' (' + subject.OU + ')';
+    };
+
+    return Promise.all([
+        readEntry('manifest.json', EfficientTextWriter),
+        readEntry('META-INF/cose.sig', Uint8ArrayWriter),
+        readEntry('META-INF/mozilla.rsa', Uint8ArrayWriter),
+    ]).then(function(entries) {
+        var manifest_json = entries[0];
+        var cose_sig = entries[1];
+        var pkcs7_sig = entries[2];
+
+        if (cose_sig) {
```

Comment:
You should first parse the subject, and if `subject.CN` is not [a valid extension ID](https://searchfox.org/mozilla-central/rev/2738efcf98d746c8810819e7c0634b2c57212a8d/toolkit/components/extensions/schemas/manifest.json#421-431), fall back to the one in manifest.json.
When the extension ID exceeds 64 characters, the certificate does not contain the ID, but a hash of it: https://searchfox.org/mozilla-central/rev/b3fd653bc6078b3be4a8d06db39eddc5714755da/toolkit/mozapps/extensions/internal/XPIInstall.jsm#802-809

Reply:
Oh, I was under the impression that the UUID was already such a hashed form... I'll have a look.

Reply:
Oh, checking cose.sig and mozilla.rsa was a deliberate decision because:
```
        // manifest.json is usually checked first, but it does not contain the
        // certification source (e.g. Production), so check it after others.
```

I should try to combine the extension ID info from manifest.json with other sources I guess.